co-simulation/lib/mosaic-common-utils/src/main/java/org/eclipse/mosaic/lib/CommonUtil/ambassador/CommonMessageAmbassador.java

```diff
@@ -0,0 +1,349 @@
+    private CarmaV2xMessageReceiver v2xMessageReceiver;
+    private Thread v2xRxBackgroundThread;
+    protected M commonInstanceManager = (M) new CommonInstanceManager();
+    protected int timeSyncSeq = 0;
+    protected Class<T> messageClass;
+    protected Class<C> configClass;
+    
+    @Override
+    public boolean isTimeRegulating() {
+        return true;
+    }
+
+    @Override
+    public boolean isTimeConstrained() {
+        return true;
+    }
+
+    public CommonMessageAmbassador(AmbassadorParameter ambassadorParameter, Class<T> messageClass, Class<C> configClass) {
+        super(ambassadorParameter);
+        this.messageClass = messageClass;
+        this.configClass = configClass;
+        try {
+            // Read the CARMA message ambassador configuration file
+            commonConfiguration = new ObjectInstantiation<>(configClass, log)
+                    .readFile(ambassadorParameter.configuration);
+        } catch (InstantiationException e) {
+            log.error("Configuration object could not be instantiated: ", e);
+        }
+
+        log.info("The update interval of Common message ambassador is " + commonConfiguration.updateInterval + " .");
+
+        // Check the CARMA update interval
+        if (commonConfiguration.updateInterval <= 0) {
+            throw new RuntimeException("Invalid update interval for CARMA message ambassador, should be >0.");
+        }
+        log.info("Common message ambassador is generated.");
+    }
+
+    @Override
+    public void initialize(long startTime, long endTime) throws InternalFederateException {
+        super.initialize(startTime, endTime);
+
+        currentSimulationTime = startTime;
+        try {
+            rti.requestAdvanceTime(currentSimulationTime, 0, (byte) 1);
+        } catch (IllegalValueException e) {
+            log.error("Error during advanceTime request", e);
+            throw new InternalFederateException(e);
+        }
+
+        // Initialize listener socket and thread for Common Registration messages
+        commonRegistrationReceiver = (R) new CommonRegistrationReceiver(messageClass);
+        commonRegistrationReceiver.init();
+        registrationRxBackgroundThread = new Thread(commonRegistrationReceiver);
+        registrationRxBackgroundThread.start();
+
+        // Initialize listener socket and thread for Common NS-3 Adapter messages
+        v2xMessageReceiver = new CarmaV2xMessageReceiver();
+        v2xMessageReceiver.init();
+        v2xRxBackgroundThread = new Thread(v2xMessageReceiver);
+        v2xRxBackgroundThread.start();
+    }
+
+    @Override
+    public synchronized void processTimeAdvanceGrant(long time) throws InternalFederateException {
+
+        if (time < currentSimulationTime) {
+            // process time advance only if time is equal or greater than the next
+            // simulation time step
+            return;
+        }
+        log.info("Common message ambassador processing timestep to {}.", time);
```

Comment:
At runtime, will the formatting for this logger show the actual inheriting class or will all log messages now seem to come form "CommonMessageAmbassador"? Might be confusing to wade through if all the CARMA ambassador's log statements get mixed together, worth considering if we can add the subclass name to these log statements somehow. There's got to be a way. My java's a bit rusty so I'm not sure what `this.getClass()` would return here, whether that'd give the superclass or the subclass, but something along those lines has to work.

Reply:
I actually tested with this.getClass(), the only problem I potentially see it's maybe it too long and complicate for a log, so I use this.getClass().getSimpleName() instead, it ouputs the class name but not the package name.

Reply:
Yeah, I remember that method now. getSimpleName() should work great for what we're looking for. I just wasn't sure whether it'd return "CommonMessengerAmbassador" or the name of the actual subclass, but if it does actually return the  subclass name then that's exactly what we'd want!